src/language/ast.ts

```diff
@@ -360,8 +375,26 @@ export interface FieldNode {
   readonly arguments?: ReadonlyArray<ArgumentNode>;
   readonly directives?: ReadonlyArray<DirectiveNode>;
   readonly selectionSet?: SelectionSetNode;
+  // Note: Client Controlled Nullability is experimental
+  // and may be changed or removed in the future.
+  readonly required?: NullabilityModifierNode;
+}
+
+export interface RequiredModifierNode {
+  readonly kind: Kind.REQUIRED_DESIGNATOR;
+  readonly loc?: Location;
+}
+
+export interface OptionalModifierNode {
+  readonly kind: Kind.OPTIONAL_DESIGNATOR;
+  readonly loc?: Location;
 }
 
+// modifiers can be ! or ?
+export type NullabilityModifierNode =
+  | RequiredModifierNode
+  | OptionalModifierNode;
+
```

Comment:
@twof Did you drop support for changing nullability of array items?

Reply:
Not officially. I'd interpreted [your comment](https://github.com/graphql/graphql-js/pull/3281#issuecomment-987855321) as you asking for only `!` and `?` to be in this PR. Should I pull the array syntax over too?

Reply:
I'd also be open to merging in this first, so array syntax can be reviewed in a separate PR. The diff created by introducing the array syntax will be quite a bit larger than this one, and I don't want to make you review too much code at once if that would be helpful.

Reply:
@IvanGoncharov I've prepared a list syntax PR, so I can merge that into this branch if you tell me that's what you want: https://github.com/twof/graphql-js/pull/2

It actually wasn't as big as I thought it would be.

Reply:
@twof Technically it may be easier to review them separately, but practically it makes hard to understand some of the reasoning in the code. 
For example, without list syntax we don't really need 2 new node kinds in AST we can just have enum with 3 values as in your original proposal.

> Not officially. I'd interpreted your comment as you asking for only ! and ? to be in this PR. Should I pull the array syntax over too?

Sorry, about that. I would try to communicate my suggestion more cleanly.

Reply:
@IvanGoncharov All good! I've merged in the list syntax.